### PR TITLE
Mark 403s from envoy in incoming permissions way - version 2

### PR DESCRIPTION
**Pull request by slonka**
Alternative to https://github.com/allegro/envoy-control/pull/243. In this PR we simplify logic in lua scripts at the cost of adding complexity to RBACFactory. Let me know which version do you preffer and we'll merge only one.

### CODE REVIEW

File: envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua [hunk 1]
```diff
@@ -62,36 +62,40 @@ end
...
-    if (not is_shadow_denied) and allowed_client == true then
-        return true
+    if is_shadow_denied then
+        local lua_metadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.lua") or {}
+        local upstream_request_time = handle:headers():get("x-envoy-upstream-service-time")
+        local rbac_action = "shadow_denied"
+        if upstream_request_time == nil then
+            rbac_action = "denied"
+        end
+        log_request(lua_metadata, handle, rbac_action, allowed_client)
```
Comment (lukidzi) [inline envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua:74]: Looks like `allowed_client` isn't defined

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactoryTest.kt [hunk 2]
```diff
@@ -710,6 +711,7 @@ internal class RBACFilterFactoryTest {
     }
 
     @Test
+    @Disabled
```
Comment (lukidzi) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactoryTest.kt:714]: ?

File: envoy-control-tests/src/main/resources/lua_spec/ingress_rbac_logging_spec.lua [hunk 3]
```diff
@@ -428,15 +504,27 @@ describe("envoy_on_response:", function()
 
         it("with globally allowed client", function ()
             -- given
-            metadata['envoy.filters.http.rbac']['shadow_engine_result'] = 'allowed'
+            metadata['envoy.filters.http.rbac']['shadow_engine_result'] = 'denied'
```
Comment (lukidzi) [inline envoy-control-tests/src/main/resources/lua_spec/ingress_rbac_logging_spec.lua:507]: Why did you change engine result? We should log requests for allowed client (allowed,denied).

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 4]
```diff
@@ -114,25 +114,30 @@ class RBACFilterFactory(
...
-        val actualPolicies = commonPolicies + allowUnlistedPolicies
+        val shadowPolicies = (commonPolicies + loggedEndpointsPolicies)
+            .map { (endpoint, policy) -> endpoint to policy.build() }.toMap()
+        val shadowRules = RBAC.newBuilder()
+            .setAction(RBAC.Action.ALLOW)
+            .putAllPolicies(shadowPolicies)
+            .build()
+        // build needs to be called before any modifications happen
 
+        val actualPolicies = (statusRoutePolicy + restrictedEndpointsPolicies.mapValues {
```
Comment (lukidzi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:125]: Could you extract what is going with `restrictedEndpointsPolicies` to method?

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactoryTest.kt (hunk 2 above)
Comment (slonka) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactoryTest.kt:714]: Yes, this has to be fixed. I didn't want to spend time on this if the other version was chosen.

File: envoy-control-tests/src/main/resources/lua_spec/ingress_rbac_logging_spec.lua (hunk 3 above)
Comment (slonka) [reply envoy-control-tests/src/main/resources/lua_spec/ingress_rbac_logging_spec.lua:507]: > Why did you change engine result?

No, this is an error in the unit test. The result is denied on master. See this comment: https://github.com/allegro/envoy-control/pull/243#discussion_r563514380

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt (hunk 4 above)
Comment (pzmi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:125]: Or at least to a variable.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 5]
```diff
@@ -114,25 +114,36 @@ class RBACFilterFactory(
             loggedEndpointsPolicies.values
         )
 
-        val shadowPolicies = commonPolicies + loggedEndpointsPolicies
-        val actualPolicies = commonPolicies + allowUnlistedPolicies
+        val shadowPolicies = (commonPolicies + loggedEndpointsPolicies)
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:117]: I may be missing something but isn't the other version just a workaround for invalid classification of `allowed_client`? This version looks actually simpler to me but as I said, I may be missing something.

Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:117]: This method looks clearer to me like this:
```kotlin
    private fun getRules(
        incomingPermissions: Incoming,
        snapshot: GlobalSnapshot,
        roles: List<Role>
    ): Rules {

        val incomingEndpointsPolicies = getIncomingEndpointPolicies(
                incomingPermissions, snapshot, roles
        )

        val restrictedEndpointsPolicies = incomingEndpointsPolicies.asSequence()
            .filter { it.endpoint.unlistedClientsPolicy == Incoming.UnlistedPolicy.BLOCKANDLOG }
            .map { (endpoint, policy) -> "$endpoint" to policy }.toMap()

        val loggedEndpointsPolicies = incomingEndpointsPolicies.asSequence()
            .filter { it.endpoint.unlistedClientsPolicy == Incoming.UnlistedPolicy.LOG }
            .map { (endpoint, policy) -> "$endpoint" to policy }.toMap()

        val allowUnlistedPolicies = unlistedAndLoggedEndpointsPolicies(
            incomingPermissions,
            (statusRoutePolicy + restrictedEndpointsPolicies).values,
            loggedEndpointsPolicies.values
        )

        val restrictedEndpointsPoliciesWithFullAccessClient = restrictedEndpointsPolicies.mapValues {
            addFullAccessClients(it.value)
        }

        val shadowPolicies = (statusRoutePolicy + restrictedEndpointsPolicies + loggedEndpointsPolicies)
            .map { (endpoint, policy) -> endpoint to policy.build() }.toMap()
        val actualPolicies = (statusRoutePolicy + restrictedEndpointsPoliciesWithFullAccessClient + allowUnlistedPolicies)
            .map { (endpoint, policy) -> endpoint to policy.build() }.toMap()

        val actualRules = RBAC.newBuilder()
            .setAction(RBAC.Action.ALLOW)
            .putAllPolicies(actualPolicies)
            .build()

        val shadowRules = RBAC.newBuilder()
            .setAction(RBAC.Action.ALLOW)
            .putAllPolicies(shadowPolicies)
            .build()
        // build needs to be called before any modifications happen so we do not have to clone it

        return Rules(shadowRules = shadowRules, actualRules = actualRules)
    }
```
A common policy is no longer common and IMO explicitly stating what policies are going where is more readable. Also, I moved `builds` around so they diff better from a previous version.

**Review by pzmi — APPROVED** [review]
I've suggested some changes but I prefer this version :)

**Review by lukidzi — DISMISSED** [review]

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 6]
```diff
@@ -114,25 +114,36 @@ class RBACFilterFactory(
...
+            .map { (endpoint, policy) -> endpoint to policy.build() }.toMap()
+        val shadowRules = RBAC.newBuilder()
+            .setAction(RBAC.Action.ALLOW)
+            .putAllPolicies(shadowPolicies)
+            .build()
+        // build needs to be called before any modifications happen so we do not have to clone it
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt]: I do not like this but it's required.

**Review by lukidzi — APPROVED** [review]
